### ISSUE TITLE
LoRA is not injected on the conditional branch alone?

### ISSUE THREAD

**Issue by va1bhavagrawal** [Xiaojiu-z/EasyControl#92]
This is a diagram from the project website.

<img width="887" height="490" alt="Image" src="https://github.com/user-attachments/assets/0fa761de-0fe6-41b0-9763-e6b416bf1685" />

It says that the LoRA is injected only on the conditional tokens. However, the implementation seems to apply the LoRA on all the `hidden_states`, irrespective of whether they belong to the noise branch or the condition branch.

<img width="1033" height="853" alt="Image" src="https://github.com/user-attachments/assets/3ead602c-5e15-4a0c-9304-eae650ba8554" />

Is this a bug?

**Comment by Xiaojiu-z** [comment]
> This is a diagram from the project website.
> 
> <img alt="Image" width="887" height="490" src="https://private-user-images.githubusercontent.com/102590952/488818205-0fa761de-0fe6-41b0-9763-e6b416bf1685.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA3NzYzNTYsIm5iZiI6MTc2MDc3NjA1NiwicGF0aCI6Ii8xMDI1OTA5NTIvNDg4ODE4MjA1LTBmYTc2MWRlLTBmZTYtNDFiMC05NzYzLWU2YjQxNmJmMTY4NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAxOFQwODI3MzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMmJiZmIzOTY1OTBmNWQzMGFmYWQ0NzhlOGE4MTc3YWYxYzExZWY3MjUyNjYzZGFmYTI2NjIzZGRjZjE1NjcwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.wzG5TYGIALCFF9navIYhdsE1R_lg4v3zKdYAA3o9aI0">
> It says that the LoRA is injected only on the conditional tokens. However, the implementation seems to apply the LoRA on all the `hidden_states`, irrespective of whether they belong to the noise branch or the condition branch.
> 
> <img alt="Image" width="1033" height="853" src="https://private-user-images.githubusercontent.com/102590952/488818790-3ead602c-5e15-4a0c-9304-eae650ba8554.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA3NzYzNTYsIm5iZiI6MTc2MDc3NjA1NiwicGF0aCI6Ii8xMDI1OTA5NTIvNDg4ODE4NzkwLTNlYWQ2MDJjLTVlMTUtNGEwYy05MzA0LWVhZTY1MGJhODU1NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAxOFQwODI3MzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZmE5NGE1MmJhNDdhYjE3MjFlNzlhNWY0ZTRmZjQwNjVmZTFlMTkzNTY3YzRhNjVkNWFmY2E2ZDJhZjZiZDc1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2lKbcWB-BsU3EwekGmyuX28FhXYWw77bjG6NxHi71z0">
> Is this a bug?

We use mask to isolated condition tokens and noise tokens.

**Comment by xinyu-pu** [comment]
Hi, thanks for the great work! I have the same question regarding how the LoRA is applied when `hidden_states` contains multiple conditions. 

Assume that `hidden_states` $$= [C_0; C_1]$$, where $$C_0$$ and $$C_1$$ are tokens from two different conditions.

My current understanding is that the **Q** projection should work like this:

```python
attn.to_q(hidden_states)
+ [ lora_weights[0] * q_loras[0](C_0);
    lora_weights[1] * q_loras[1](C_1) ]
```

However, from the code, it looks like the implementation is doing:

```python
attn.to_q(hidden_states)
+ lora_weights[0] * q_loras[0]([C_0; C_1])
+ lora_weights[1] * q_loras[1]([C_0; C_1])
```

So each LoRA seems to be applied over the entire concatenated `hidden state` $$[C_0; C_1]$$, instead of being isolated to its own condition.

Did I misunderstand the intended behavior, or is this the expected design?

**Comment by xinyu-pu** [comment]
> Hi, thanks for the great work! I have the same question regarding how the LoRA is applied when `hidden_states` contains multiple conditions.
> 
> Assume that `hidden_states` = [ C 0 ; C 1 ] , where C 0 and C 1 are tokens from two different conditions.
> 
> My current understanding is that the **Q** projection should work like this:
> 
> attn.to_q(hidden_states)
> + [ lora_weights[0] * q_loras[0](C_0);
>     lora_weights[1] * q_loras[1](C_1) ]
> However, from the code, it looks like the implementation is doing:
> 
> attn.to_q(hidden_states)
> + lora_weights[0] * q_loras[0]([C_0; C_1])
> + lora_weights[1] * q_loras[1]([C_0; C_1])
> So each LoRA seems to be applied over the entire concatenated `hidden state` [ C 0 ; C 1 ] , instead of being isolated to its own condition.
> 
> Did I misunderstand the intended behavior, or is this the expected design?

I figured out the behavior after reading the implementation more carefully.
The masking logic inside LoRALinearLayer explains how each LoRA branch only sees its own condition block in hidden_states, 
```python
mask = torch.ones(shape, device=hidden_states.device, dtype=dtype)
mask[:, :block_size + self.number * cond_size, :] = 0
mask[:, block_size + (self.number + 1) * cond_size:, :] = 0

hidden_states = mask * hidden_states
```
So even though the conditions are concatenated, each branch is masked to isolate its own slice.

Thanks again for the great work!
